tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,754 @@
+                ffi::tiledb_query_condition_op_t_TILEDB_NOT_IN
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub enum NullnessOp {
+    IsNull,
+    NotNull,
+}
+
+impl NullnessOp {
+    pub(crate) fn capi_enum(&self) -> ffi::tiledb_query_condition_op_t {
+        match self {
+            NullnessOp::IsNull => ffi::tiledb_query_condition_op_t_TILEDB_EQ,
+            NullnessOp::NotNull => ffi::tiledb_query_condition_op_t_TILEDB_NE,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub enum CombineOp {
+    And,
+    Or,
+}
+
+impl CombineOp {
+    pub(crate) fn capi_enum(
+        &self,
+    ) -> ffi::tiledb_query_condition_combination_op_t {
+        match self {
+            CombineOp::And => {
+                ffi::tiledb_query_condition_combination_op_t_TILEDB_AND
+            }
+            CombineOp::Or => {
+                ffi::tiledb_query_condition_combination_op_t_TILEDB_OR
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum CmpValue {
+    UInt8(u8),
+    UInt16(u16),
+    UInt32(u32),
+    UInt64(u64),
+    Int8(i8),
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    Float32(f32),
+    Float64(f64),
+    String(String),
+}
+
+impl CmpValue {
+    fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            Self::UInt8(val) => val.to_le_bytes().to_vec(),
+            Self::UInt16(val) => val.to_le_bytes().to_vec(),
+            Self::UInt32(val) => val.to_le_bytes().to_vec(),
+            Self::UInt64(val) => val.to_le_bytes().to_vec(),
+            Self::Int8(val) => val.to_le_bytes().to_vec(),
+            Self::Int16(val) => val.to_le_bytes().to_vec(),
+            Self::Int32(val) => val.to_le_bytes().to_vec(),
+            Self::Int64(val) => val.to_le_bytes().to_vec(),
+            Self::Float32(val) => val.to_le_bytes().to_vec(),
+            Self::Float64(val) => val.to_le_bytes().to_vec(),
+            Self::String(val) => val.as_bytes().to_vec(),
+        }
+    }
+}
+
+macro_rules! cmp_value_impl {
+    ($ty:ty, $constructor:expr) => {
+        impl From<$ty> for CmpValue {
+            fn from(value: $ty) -> CmpValue {
+                $constructor(value)
+            }
+        }
+    };
+}
+
+cmp_value_impl!(u8, CmpValue::UInt8);
+cmp_value_impl!(u16, CmpValue::UInt16);
+cmp_value_impl!(u32, CmpValue::UInt32);
+cmp_value_impl!(u64, CmpValue::UInt64);
+cmp_value_impl!(i8, CmpValue::Int8);
+cmp_value_impl!(i16, CmpValue::Int16);
+cmp_value_impl!(i32, CmpValue::Int32);
+cmp_value_impl!(i64, CmpValue::Int64);
+cmp_value_impl!(f32, CmpValue::Float32);
+cmp_value_impl!(f64, CmpValue::Float64);
+cmp_value_impl!(String, CmpValue::String);
+
+impl From<&str> for CmpValue {
+    fn from(val: &str) -> CmpValue {
+        CmpValue::String(val.to_string())
+    }
+}
+
+// N.B. I initially tried slices here, but that breaks the Deserialize trait.
```

Comment:
Right, I initially tried using `&[T]` but that caused the Deserializer trait to complain that arbitrary length slices aren't supported. I assumed it was an ownership thing because during deserialization who would own the slices?